types/react/jsx-dev-runtime.d.ts

```diff
@@ -11,3 +12,34 @@ export namespace JSX {
     interface IntrinsicClassAttributes<T> extends React.JSX.IntrinsicClassAttributes<T> {}
     interface IntrinsicElements extends React.JSX.IntrinsicElements {}
 }
+
+export interface JSXSource {
+    /**
+     * The source file where the element originates from.
+     */
+    fileName?: string;
+
+    /**
+     * The line number where the element was created.
+     */
+    lineNumber?: number;
+
+    /**
+     * The column number where the element was created.
+     */
+    columnNumber?: number;
+}
```

Comment:
It depends on the compiler. It’s not really properly specified. Perhaps it should allow both an unspecified property or an `undefined` value. In this case it may be better to allow more loose than stricter types.

Reply:
If someone chooses a strict compiler, and Babel or whatnot passed `fileName: undefined`, we’d still want the types to be correct right?
I quickly glanced through their code and they always pass the properties.
From the MDX side, I definitely know we sometimes pass `undefined` as a value for line number and col number, with the fields defined. https://github.com/syntax-tree/estree-util-build-jsx/blob/1538244b56aac303293cafd831bb9e602e57f75c/lib/index.js#L417